agent.go

```diff
@@ -0,0 +1,157 @@
+	// Default value whether the agent is enabled or not.
+	defaultEnabled = true
+	// Default value for batch of data sent per flush interval.
+	defaultBatchSize = 10000
+	// Default size of internal buffers beyond which received data is dropped
+	defaultMaxBufferSize = 50000
+	// Default interval (in seconds) at which to flush data to Wavefront.
+	defaultFlushIntervalSeconds = 1
+)
+
+// NewWavefrontAgent returns a new agent.
+func NewWavefrontAgent(w *WavefrontConfig) *WavefrontAgent {
+	// Create a new instance of the WavefrontAgent.
+	wfAgent := &WavefrontAgent{
+		metrics:  make(map[string]float64),
+		counters: make(map[string]float64),
+	}
+
+	// Create an empty map of point tags if no tags exist yet.
+	if w.PointTags == nil {
+		w.PointTags = make(map[string]string)
+	}
+
+	// Create the configuration to connect to Wavefront. Details are gathered from both
+	// the WavefrontConfig and the environment variables. If both WavefrontConfig and
+	// environment variables have a value for a specific setting, the environment variable
+	// takes precedence.
+	enabled := &defaultEnabled
+	envEnabled := os.Getenv("WAVEFRONT_ENABLED")
+	if w.Enabled != nil {
+		enabled = w.Enabled
+	}
+	if envEnabled != "" {
+		enabled = stringToBool(envEnabled)
+	}
+
+	envServer := os.Getenv("WAVEFRONT_URL")
+	server := &envServer
+	if w.Server != nil && len(envServer) == 0 {
+		server = w.Server
+	}
+
+	envToken := os.Getenv("WAVEFRONT_API_TOKEN")
+	token := &envToken
+	if w.Token != nil && len(envToken) == 0 {
+		token = w.Token
+	}
+
+	batchSize := &defaultBatchSize
+	envBatchSize := os.Getenv("WAVEFRONT_BATCH_SIZE")
+	if w.BatchSize != nil {
+		batchSize = w.BatchSize
+	}
+	if envBatchSize != "" {
+		batchSizeInt, err := stringToInt(envBatchSize)
+		if err == nil {
+			w.BatchSize = batchSizeInt
+			batchSize = batchSizeInt
+		}
+	}
+
+	maxBufferSize := &defaultMaxBufferSize
+	envMaxBufferSize := os.Getenv("WAVEFRONT_MAX_BUFFER_SIZE")
+	if w.MaxBufferSize != nil {
+		maxBufferSize = w.MaxBufferSize
+	}
+	if envMaxBufferSize != "" {
+		maxBufferSizeInt, err := stringToInt(envMaxBufferSize)
+		if err == nil {
+			w.MaxBufferSize = maxBufferSizeInt
+			maxBufferSize = maxBufferSizeInt
+		}
+	}
+
+	dc := &wavefront.DirectConfiguration{
+		Server:               *server,
+		Token:                *token,
+		BatchSize:            *batchSize,
+		MaxBufferSize:        *maxBufferSize,
+		FlushIntervalSeconds: 1,
+	}
+
+	sender, err := wavefront.NewDirectSender(dc)
+	if err != nil {
+		log.Printf("ERROR :: %s", err.Error())
+	}
+
+	wfAgent.sender = sender
+	wfAgent.WavefrontConfig = w
+	wfAgent.WavefrontConfig.Enabled = enabled
+
+	return wfAgent
+}
+
+// WrapHandler wraps the handler
+func (wa *WavefrontAgent) WrapHandler(handler interface{}) interface{} {
+	if !*wa.Enabled {
+		return handler
+	}
+
+	return wrapHandler(handler, wa)
+}
+
+// RegisterMetric adds a new metric to be sent to Wavefront
+func (wa *WavefrontAgent) RegisterMetric(name string, value float64) {
```

Comment:
I remember discussing on the lines of exposing all "send methods in wavefront-sdk-go" via the agent wrapper.
The Wrapper should cover APIs like : https://github.com/wavefrontHQ/wavefront-sdk-go/blob/86329984eb326ae0dc2b51ff12944fb4f9c70638/senders/direct.go#L83-L136

We should provide the user to send any type of metric that wavefront-sdk-go supports.

Reply:
The way I see it, there are two routes we can take to accomplish this:

1. Make the `sender` object in the WavefrontAgent config exported so it's accessible in the code and devs can "just" call the methods they need.
2. Add wrappers for each of the methods that encapsulate the methods from the Wavefront sender.

The first option is the easier one but would tie this lib a little more closely to the Wavefront SDK. The second one is a little nicer but would increase the overall code complexity. Any preference?

Reply:
We should avoid option #1. Changes to tier1 wavefront sdk (wavefront-sdk-go) shouldn't cause the user lambda code to change. We should provide them with stubs in your agent i.e option #2.